tough/src/http.rs

```diff
@@ -180,81 +188,217 @@ impl RetryState {
-                        return Err(err).context(error::HttpFetch { url: url.clone() });
-                    }
+            HttpResult::Fatal(err) => {
+                trace!("{:?} - returning fatal error from fetch: {}", r, err);
+                return Err(http_error::HttpError::FetchFatal { source: err });
+            }
+            HttpResult::FileNotFound(err) => {
+                trace!("{:?} - returning file not found from fetch: {}", r, err);
+                return Err(http_error::HttpError::FetchFileNotFound { source: err });
+            }
+            HttpResult::Retryable(err) => {
+                trace!("{:?} - retryable error: {}", r, err);
+                if r.current_try >= cs.tries - 1 {
+                    debug!("{:?} - returning failure, no more retries: {}", r, err);
+                    return Err(http_error::HttpError::FetchNoMoreRetries {
+                        tries: cs.tries,
+                        source: err,
+                    });
                 }
-                // we will retry if possible, otherwise we will return this err.
-                err
             }
-        };
-
-        // increment the retry state and continue trying unless we are out of tries
-        if r.current_try >= cs.tries - 1 {
-            return Err(retry_err).context(error::HttpRetries {
-                url: url.clone(),
-                tries: cs.tries,
-            });
         }
+
         r.increment(&cs);
         std::thread::sleep(r.wait);
     }
 }
 
-fn build_request(client: &Client, next_byte: usize, url: &Url) -> Result<Request> {
+/// Much of the complexity in the `fetch_with_retries` function is in deciphering the `Result`
+/// we get from `reqwest::Client::execute`. Using this enum we categorize the states of the
+/// `Result` into the categories that we need to understand.
+enum HttpResult {
+    /// We got a response with an HTTP code that indicates success.
+    Ok(reqwest::blocking::Response),
+    /// We got an `Error` (other than file-not-found) which we will not retry.
+    Fatal(reqwest::Error),
+    /// The file could not be found (HTTP status 403 or 404).
+    FileNotFound(reqwest::Error),
+    /// We received an `Error`, or we received an HTTP response code that we can retry.
+    Retryable(reqwest::Error),
+}
+
+/// Takes the `Result` type from `reqwest::Client::execute`, and categorizes it into an
+/// `HttpResult` variant.
+impl Into<HttpResult> for Result<reqwest::blocking::Response, reqwest::Error> {
+    fn into(self) -> HttpResult {
+        match self {
+            Ok(response) => {
+                trace!("response received");
+                // checks the status code of the response for errors
+                parse_response_code(response)
+            }
+            Err(err) => {
+                // an error occurred before the HTTP header could be read
+                trace!("retryable error during fetch: {}", err);
+                HttpResult::Retryable(err)
```

Comment:
FIXME: These are not server errors, these are any errors that occur before an HTTP status is comprehended by reqwest. For example
- Broken pipe before the status is reached
- connection timeout
- reqwest exploded

I'll come back to this and see how messy it would be to make a decision based on the error here.

Your confusion is understandable. The comprehension of these errors is one of the most confusing parts of implementing this module.

Reply:
This is improved [f5bfcbb](https://github.com/awslabs/tough/compare/55ff0282802e8c7829762f923d6e92ced3a49b93..f5bfcbb4be7dcafa0a25b4ff8fb02649a2c886f1)